# 📜 Project: Data Understanding Report

 Phase 1: Data Exploration & Initial Assessment

In [66]:

import pandas as pd 
import numpy as np

df = pd.read_csv("ts_transport_registration_01_01_2025to31_01_2025.csv")

### Step 1: Data Understanding
### What columns are present?

In [67]:
df.info()
df.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 343137 entries, 0 to 343136
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   slno            343137 non-null  int64  
 1   registrationNo  341450 non-null  object 
 2   reserve_no      23835 non-null   object 
 3   regvalidfrom    341445 non-null  object 
 4   regvalidto      343137 non-null  object 
 5   makerName       343137 non-null  object 
 6   modelDesc       343137 non-null  object 
 7   bodyType        343133 non-null  object 
 8   cc              343137 non-null  float64
 9   cylinder        343137 non-null  int64  
 10  fuel            339359 non-null  object 
 11  hp              343137 non-null  float64
 12  seatCapacity    343137 non-null  int64  
 13  OfficeCd        343137 non-null  object 
 14  fromdate        343137 non-null  object 
 15  todate          343137 non-null  object 
dtypes: float64(2), int64(3), object(11)
memory usage: 41.9+ 

(343137, 16)

In [68]:
#Contains the below columns:
### slno: serial_number
### Regn_No: Registration Number
##  Reserve_no :- vechile pre registration number 
### Regn_VFDt: Registration Valid From Date 
### Regn_VTDt: Registration Valid To Date 
### Maker_Name: Automobile Maker Name Model_Desc: 
### Vehicle Model Description BodyTyp: Vehicle Body 
### Type CC:Engine Capacity
### Cylinder: Number of Cylinders in the Vehicle 
### Engine Fuel: Type of Fuel 
### HP: Horse of the Vehicle 
### Seat_Capacity: Seat capacity of the vehicle 
### Apprved_Dt: Approval Date by RTA 
### OfficeCd: Registration Office



### What kind of data does each column contain?

In [69]:
df.dtypes

slno                int64
registrationNo     object
reserve_no         object
regvalidfrom       object
regvalidto         object
makerName          object
modelDesc          object
bodyType           object
cc                float64
cylinder            int64
fuel               object
hp                float64
seatCapacity        int64
OfficeCd           object
fromdate           object
todate             object
dtype: object

###  Are there any missing or unusual values?


In [70]:
df.isna().sum()

slno                   0
registrationNo      1687
reserve_no        319302
regvalidfrom        1692
regvalidto             0
makerName              0
modelDesc              0
bodyType               4
cc                     0
cylinder               0
fuel                3778
hp                     0
seatCapacity           0
OfficeCd               0
fromdate               0
todate                 0
dtype: int64

In [71]:
##  What You Need to Do in This Step ?

### 2. View sample data Use .head() to preview the first few records.



In [72]:
df.head()

slno registrationNo reserve_no regvalidfrom  \
0   9287073      TS36G2115        NaN   02/03/2020   
1  10311462      TS21J6061        NaN   29/01/2021   
2  10949878     TS08HW0365        NaN   27/11/2021   
3   5255752     TS08FB0182        NaN   03/11/2016   
4   1506087     AP29AJ8075        NaN   25/11/2009   

                      regvalidto                       makerName  \
0  2035-03-01T17:12:20.747+05:30       MARUTI SUZUKI INDIA LTD.,   
1  2036-01-28T16:46:36.317+05:30       MARUTI SUZUKI INDIA LTD.,   
2  2036-11-26T16:32:35.737+05:30  HONDA MOTORCYCLE&SCOOTER(I)P L   
3   2031-11-02T16:55:19.72+05:30  HONDA MOTORCYCLE&SCOOTER(I)P L   
4  2030-01-01T11:00:52.313+05:30       MARUTI SUZUKI INDIA LTD.,   

                                         modelDesc   bodyType      cc  \
0               MARUTI XL6 SMART HYBRID ALPHA BSVI  HATCHBACK  1462.0   
1                         MARUTI S-PRESSO VXI BSVI  HATCHBACK   998.0   
2               ACTIVA 6G WITH LED HEAD LIGHT BSVI       Solo   109.0   
3  ACTIVA 3G W ELECTRIC A START&K START &CBS BSIII       Solo   109.0   
4                            MARUTI  RITZ LDI BSIV       Solo  1248.0   

   cylinder    fuel      hp  seatCapacity      OfficeCd    fromdate  \
0         4  PETROL  103.25             6  RTA SIDDIPET  01/01/2025   
1         3  PETROL   67.05             5  RTA JAGITYAL  01/01/2025   
2         1  PETROL    7.68             2     RTA UPPAL  01/01/2025   
3         1  PETROL    8.00             2     RTA UPPAL  01/01/2025   
4         4  DIESEL   55.00             5  RTA SIDDIPET  01/01/2025   

       todate  
0  02/01/2025  
1  02/01/2025  
2  02/01/2025  
3  02/01/2025  
4  02/01/2025

In [73]:
df.describe(include = "all")

slno registrationNo reserve_no regvalidfrom  \
count   3.431370e+05         341450      23835       341445   
unique           NaN         174057      12190         6572   
top              NaN      KA36C7339  TG08M4995   10/01/2025   
freq             NaN              2          4         9658   
mean    1.114881e+07            NaN        NaN          NaN   
std     4.294076e+06            NaN        NaN          NaN   
min     2.570000e+02            NaN        NaN          NaN   
25%     8.481106e+06            NaN        NaN          NaN   
50%     1.421286e+07            NaN        NaN          NaN   
75%     1.427993e+07            NaN        NaN          NaN   
max     1.436026e+07            NaN        NaN          NaN   

                           regvalidto                       makerName  \
count                          343137                          343137   
unique                          71783                             748   
top     2040-01-03T18:51:46.993+05:30  HONDA MOTORCYCLE&SCOOTER(I)P L   
freq                              942                           62617   
mean                              NaN                             NaN   
std                               NaN                             NaN   
min                               NaN                             NaN   
25%                               NaN                             NaN   
50%                               NaN                             NaN   
75%                               NaN                             NaN   
max                               NaN                             NaN   

                  modelDesc bodyType             cc       cylinder    fuel  \
count                343137   343133  343137.000000  343137.000000  339359   
unique                10029       39            NaN            NaN      13   
top     ACTIVA-DLX BSVI-PH2     Solo            NaN            NaN  PETROL   
freq                  11560   181842            NaN            NaN  231422   
mean                    NaN      NaN     835.948610       3.989415     NaN   
std                     NaN      NaN    1141.992181      75.539689     NaN   
min                     NaN      NaN       0.000000       0.000000     NaN   
25%                     NaN      NaN     113.000000       1.000000     NaN   
50%                     NaN      NaN     163.000000       1.000000     NaN   
75%                     NaN      NaN    1199.000000       4.000000     NaN   
max                     NaN      NaN   29556.000000    2979.000000     NaN   

                   hp   seatCapacity        OfficeCd    fromdate      todate  
count   343137.000000  343137.000000          343137      343137      343137  
unique            NaN            NaN              56          30          30  
top               NaN            NaN  RTA RANGAREDDY  09/01/2025  10/01/2025  
freq              NaN            NaN           28675       17455       17455  
mean        47.806952       3.229427             NaN         NaN         NaN  
std        201.605959       3.251204             NaN         NaN         NaN  
min          0.000000       0.000000             NaN         NaN         NaN  
25%          8.050000       2.000000             NaN         NaN         NaN  
50%         13.810000       2.000000             NaN         NaN         NaN  
75%         73.760000       5.000000             NaN         NaN         NaN  
max      45500.000000      61.000000             NaN         NaN         NaN

In [74]:
df.isnull().sum()

slno                   0
registrationNo      1687
reserve_no        319302
regvalidfrom        1692
regvalidto             0
makerName              0
modelDesc              0
bodyType               4
cc                     0
cylinder               0
fuel                3778
hp                     0
seatCapacity           0
OfficeCd               0
fromdate               0
todate                 0
dtype: int64

### value_counts()and nuique each columnss

In [75]:
 for i in df.columns:
  print("*"*(5) ,i, "*"*(5))
  print(df[i].value_counts())
  print(df[i].nunique())

***** slno *****
slno
14334590    2
12831877    2
12572951    2
1595865     2
10091592    2
           ..
14331631    1
14350878    1
14333556    1
14340093    1
8718345     1
Name: count, Length: 174949, dtype: int64
174949
***** registrationNo *****
registrationNo
KA36C7339     2
TS13ER3987    2
AP09BD7509    2
TG36A3084     2
TS10FG4697    2
             ..
AP05BT3456    1
TS29G5206     1
AP24AS8601    1
TS29E1035     1
TS29Q0099     1
Name: count, Length: 174057, dtype: int64
174057
***** reserve_no *****
reserve_no
TG08M4995    4
TG236660     4
TG16C2889    4
TG07N5056    4
TG07N5567    4
            ..
TG09E0019    1
TG11B0279    1
TG09E0003    1
TG07P2882    1
TG07P2883    1
Name: count, Length: 12190, dtype: int64
12190
***** regvalidfrom *****
regvalidfrom
10/01/2025    9658
04/01/2025    9374
06/01/2025    9140
18/01/2025    9082
09/01/2025    8803
              ... 
18/06/1969       1
12/08/2002       1
27/08/2002       1
03/03/2011       1
02/09/1958       1
Name: count, Le

## Questions to Answer in Your Report

### What is the distribution of key fields like:

### " 1 Slno"

In [76]:
 df["slno"].duplicated(keep= "first").sum() 

np.int64(168188)

In [77]:
w = df["slno"].duplicated() == True
indices = np.where(w.values == True)[0]  # indices is a numpy array


In [78]:
sum(df["slno"].value_counts()==2)

168188

In [79]:
df["slno"].nunique()

174949

In [80]:
# Step 1: Get boolean series where value_counts == 2
w = df["slno"].value_counts() == 2

print(w.sum())

# Step 2: Count how many are exactly True (appear exactly twice)
duplicated_twice_count = w.sum()

# Step 3: Count how many are not exactly twice
not_duplicated_twice_count = len(w) - duplicated_twice_count

print("Duplicated (appear exactly twice):", duplicated_twice_count)
print("Not Duplicated (appear ≠ 2):", not_duplicated_twice_count)


168188
Duplicated (appear exactly twice): 168188
Not Duplicated (appear ≠ 2): 6761


In [81]:
vc = df["slno"].value_counts()
exactly_twice_ids = vc[vc == 2].index

# Rows where slno appears exactly twice
duplicated_twice_rows = df[df["slno"].isin(exactly_twice_ids)]

# Count rows that are part of duplicated slnos
print("Rows with slno appearing exactly twice:", len(duplicated_twice_rows))

# Total rows that are not part of those
not_duplicated_twice_rows = df[~df["slno"].isin(exactly_twice_ids)]
print("Other rows:", len(not_duplicated_twice_rows))

# Sanity check
print("Total rows:", len(df))
print("Check:", len(duplicated_twice_rows) + len(not_duplicated_twice_rows) == len(df))


Rows with slno appearing exactly twice: 336376
Other rows: 6761
Total rows: 343137
Check: True


### Single serial numbers (appear only once or not exactly twice): 6,761
###  Duplicated serial numbers (appear exactly twice): 168,188 unique

### Total rows from duplicated serials: 168,188 × 2 = 336,376  There  is no missing values
  
### total rows =336,376 + 6,761 = 343,137


## 2 registartionNO

In [82]:
df["registrationNo"].duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
343132     True
343133     True
343134    False
343135    False
343136    False
Name: registrationNo, Length: 343137, dtype: bool

In [83]:
# registraion number :
df["registrationNo"].value_counts()


registrationNo
KA36C7339     2
TS13ER3987    2
AP09BD7509    2
TG36A3084     2
TS10FG4697    2
             ..
AP05BT3456    1
TS29G5206     1
AP24AS8601    1
TS29E1035     1
TS29Q0099     1
Name: count, Length: 174057, dtype: int64

In [84]:
df["registrationNo"].nunique()

174057

In [85]:
# Step 1: Get boolean series where value_counts == 2
w = df["registrationNo"].value_counts() == 2

# Step 2: Count how many are exactly True (appear exactly twice)
duplicated_twice_count = w.sum()

# Step 3: Count how many are not exactly twice
not_duplicated_twice_count = len(w) - duplicated_twice_count

print("Duplicated (appear exactly twice):", duplicated_twice_count)
print("Not Duplicated (appear ≠ 2):", not_duplicated_twice_count)
print("total nuique numbers", duplicated_twice_count+ not_duplicated_twice_count)


Duplicated (appear exactly twice): 167393
Not Duplicated (appear ≠ 2): 6664
total nuique numbers 174057


### registrationNo 
### Duplicated registration numbers (appear exactly twice):
167,393 unique values → 167,393 × 2 = 334,786 rows
### Single registration numbers (appear only once):6,664 unique values → 6,664 rows
### Total unique registration numbers (nunique):   167,393 + 6,664 = 174,057
### Total rows in the dataset:    334,786 (duplicated) + 6,664 (single) = 341,450 without missing values 
#### 1687  are missing values.  So reserve_nno  values 92% match registration number
##### registration==null and reserve_no == notnull  we get 1686 number we fill  the registration number 
       it and remove One row  we remove it



#  3 reserve no

In [86]:
# Reserve_no

df["reserve_no"].value_counts()


reserve_no
TG08M4995    4
TG236660     4
TG16C2889    4
TG07N5056    4
TG07N5567    4
            ..
TG09E0019    1
TG11B0279    1
TG09E0003    1
TG07P2882    1
TG07P2883    1
Name: count, Length: 12190, dtype: int64

In [87]:

four  = df["reserve_no"].value_counts() == 4
three  = df["reserve_no"].value_counts() == 3
two  = df["reserve_no"].value_counts() == 2
one   = df["reserve_no"].value_counts() == 1
sum = four.sum()+three.sum() + two.sum() +one.sum()
print(sum)


#print(four.sum())





12190


In [88]:
df["reserve_no"].nunique()


12190

In [89]:
two = 11625*2
three = 3*1
four = 4* 6
one = 558 *1
total = one +two + three +four
total
print(one)
print(two)
print(three)
print(four)
print(total)
df["reserve_no"].nunique()


558
23250
3
24
23835


12190

In [90]:
df["reserve_no"].notnull().sum()


np.int64(23835)

In [91]:
# Condition: reserve_no is present but registrationNo is missing
condition = (df['reserve_no'].notnull()) & (
    df['registrationNo'].isna() | (df['registrationNo'].astype(str).str.strip() == '')
)

# Count of such rows
count_reserved_but_no_registration = condition.sum()

# Total number of rows in dataset
total_rows = len(df)

# Calculate percentage
percentage = (count_reserved_but_no_registration / total_rows) * 100

print(f"Count: {count_reserved_but_no_registration}")
print(f"Total Rows: {total_rows}")
print(f"Percentage: {percentage:.2f}%")


Count: 1686
Total Rows: 343137
Percentage: 0.49%


### reserve_no:- pre  registration numbers:
#### Out of 343,137 total rows, only 23,835 rows contain This means 317,302 rows (~93.05%) .are null values
### Repetition Frequency:
Among the non-null values:
 ### 6  appears 4 times,    6X4 = 24
 ### 1  appears 3 times,    1 X 3 = 3  
 ### 11625  appears 2 times, 11625 X 2 = 23250
  ### 558 appear only once 558 X 1 = 558 
#### VUnique Count & Match to Non-null Rows:   the  values equal to = df[“reserve_no”].unique()  12 ,190
The total number of unique reserve_no values is 13,835. 





## 3,4, 14,15 time and Date Handling
##### regvalidfrom 
This column contains 1,692 missing values this is tough to deal.
##### Regvalidfrom, regvalidto, from date, and to date  Handling:
These columns have no missing values but are also stored as object types. To enable proper date comparisons or calculations, they should be converted to datetime format.



 ## 5 makerName

In [92]:
print(df["makerName"].value_counts())

makerName
HONDA MOTORCYCLE&SCOOTER(I)P L            62617
MARUTI SUZUKI INDIA LTD.,                 38653
BAJAJ AUTO LTD                            35937
HERO MOTOCORP LTD                         28550
TVS MOTOR COMPANY LTD                     24820
                                          ...  
WASIM MEERA ENGINEERING WORKS                 1
M/S VENKATARAMANA AGRO AGENCY,METPALLY        1
NEW SRI SHIVA ENGG WORKS MBNR                 1
M/S SRI RAMA AGRO INDUSTRIES                  1
SRI SAI LAXMI TULJA ENGG WORKS                1
Name: count, Length: 748, dtype: int64


In [93]:
print(df["makerName"].nunique())

748


## makerName

- **makerName** represents the manufacturer of each registered vehicle.  
- There are **no missing values** in this column.
- There are **778 unique maker names** available.

### Top 5 Dominant Manufacturers

| Manufacturer                        | Count    |
|--------------------------------------|----------|
| ONDA MOTORCYCLE & SCOOTER (I) P L    | 62,617   |
| MARUTI SUZUKI INDIA LTD.             | 38,653   |
| BAJAJ AUTO LTD                       | 35,937   |
| HERO MOTOCORP LTD                    | 28,550   |
| TVS MOTOR COMPANY LTD                 | 24,820   |

---

### Bottom 3 Rare Entries in modelDesc

| Model Description                                 | Count |
|---------------------------------------------------|-------|
| BMW- M340I LCI BSVI-PH2                           | 1     |
| LPT 407 EX/34 BSIV FFC TC CAB LB PF               | 1     |
| CROP TIGER 40W M460 SELF PROPELLED CH BSIV        | 1     |

---

### Skewness Observation

The dataset is **highly skewed** in both `makerName` and `modelDesc`. A few manufacturers and models dominate the majority of the dataset (e.g., ONDA and ACTIVA), while thousands of other models appear only once. This skew may be due to regional vehicle popularity or high sales concentration of specific manufacturers/models in the area.

## 6 modelDesc

In [94]:
df["modelDesc"].value_counts()

modelDesc
ACTIVA-DLX BSVI-PH2                           11560
PASSION + I 3S DRS BSVI-PH2                    5484
UNICORN BSVI-PH2                               5083
ACTIVA 125 DISC BSVI-PH2                       4857
SHINE 125 DRUM BSVI-PH2                        3966
                                              ...  
BMW- M340I LCI BSVI-PH2                           1
6055 F24- (BRANDNAME-FARMTRAC) BSIIIA             1
LPT 407 EX/34 BSIV FFC TC CAB LB PF               1
CROP TIGER 40W M460 SELF PROPELLED CH BSIV        1
MARUTI S-PRESSO VXI (O) 1L ISS AGS BSVI           1
Name: count, Length: 10029, dtype: int64

In [95]:
df["modelDesc"].nunique()

10029

### modelDesc 
### The modelDesc column describes the specific vehicle models,.
### It contains no missing values   . And  798 unique 
### Popular models such as ACTIVA-DLX BSVI-PH2 (11,560 records) dominate the dataset.
### Rare models like MW- M340I LCI BSVI-PH2, 6055 F24- (BRANDNAME-FARMTRAC) BSIIIA, LPT 407 EX/34 BSIV FFC TC CAB LB ### PF, and CROP TIGER 40W M460 SELF PROPELLED CH BSIV appear only once, 




 ##  7 bodyType

In [96]:
df["bodyType"].value_counts()

bodyType
Solo                           181842
Sedan                           43885
Open                            36500
HATCHBACK                       35328
Saloon                          13386
Station Wagon                    6978
Monocoque                        6686
Hard Top                         5352
Hackney                          4384
Closed                           3101
TIPPER                           1755
TANKER                            945
COMBINED HARVESTER                784
Soft Top                          486
Coupe                             391
CONCRETE MIXER                    159
TRANSITMIXER                      154
Van                               146
Crane                             109
CAGE                              100
Tourer                             90
MOBILE CRANE                       89
Compressor Mounted                 87
PLATFORM                           85
Jeep                               74
Steel Body                         57
CHA

In [97]:
df["bodyType"].nunique()


39

### BodyType
 ### The bodyType of the vehicle's body. It helps classify vehicles such as Solo, Sedan, HATCHBACK, Van, etc.
  #### There are no missing values in this column. And  39 unique vehicle body types present in the dataset.
### Top 5 Most Frequent Body Types:
#### Solo – 181,842,   Sedan – 43,885  ,Open – 36,500,  HATCHBACK – 35,328 ,Saloon – 13,386
#### Skewness and Handling:
#### The data is highly skewed toward a few categories like 'Solo' and 'Sedan', which dominate the distribution.




## 8 cc cybic centimeters

In [98]:
df["cc"].value_counts()

cc
124.0     36870
97.0      23231
1197.0    22520
110.0     22085
125.0     21331
          ...  
147.5         1
3387.0        1
3295.0        1
3023.0        1
2878.0        1
Name: count, Length: 588, dtype: int64

#### Cc cubic Centiemeters
#### The cc column represents the engine displacement (in cubic centimeters) 
#### It has no missing values, which  and 588   unique categories.
#### Frequently occurring values like 124.0 cc, 97.0 cc, and 1197.0 cc suggest a dominance of small-engine vehicles, such as scooters and compact cars.
#### Rare engine capacities like 147.5 cc, 3387.0 cc, 3295.0 cc, and 3023.0 cc appear only once, indicating a wide variety of vehicle types, including possibly high-performance or commercial vehicles


## 9 Cylinder Column


In [99]:
df["cylinder"].value_counts()

cylinder
1       185720
4        88853
3        40469
0        17168
6         6543
2         4058
2979       221
5           72
8           26
10           4
12           3
Name: count, dtype: int64

In [100]:
df["cylinder"].nunique()

11


#### The cylinder  represent number of cylinder available to the vehicles  there is no missing values and 13  unique values
#### 1 cylinder: 185,720 Two-wheelers)
#### 4 cylinders: 88,853 (Small cars)
#### 3 cylinders: 40,469( Compact vehicles)
##### Suspicious value: 2979 (221 times) – likely a data entry error
6, 8, 10, 12 cylinders appear in low frequency



# 8 fuel

df["fuel"].value_counts()

In [101]:
df["fuel"].value_counts()

fuel
PETROL             231422
DIESEL              81638
BATTERY             12709
CNG PETROL           7156
PETROL LPG           5243
-1                    439
CNG                   374
DIESEL ELECTRIC       195
PETROL ELECTRIC       135
LPG                    32
0                      10
petrol                  4
CNG DIESEL              2
Name: count, dtype: int64

In [102]:
df["fuel"].nunique()

13

fuel column 
###  It has 3,778 missing values and  13 unique valid categories.
### and includes invalid entries like -1, 0, and lowercase petrol.
### Top 3 fuels: PETROL (231,422), DIESEL (81,638), BATTERY (12,709) — show strong data skew.
### Bottom 3 fuels: CNG DIESEL (2), petrol (4), 0 (10) — indicate typos or rare/invalid data. And -1 439 

### Missing and invalid entries can be imputed using cc, cylinder, hp, seatCapacity, and modelDesc.


# hp HorsePower

In [103]:
df["hp"].nunique()

1055

In [104]:
df["hp"].value_counts()

hp
7.91      17754
7.74      13936
8.58      10111
88.51      9968
8.19       8446
          ...  
33.80         2
570.00        1
59.12         1
56.10         1
331.00        1
Name: count, Length: 1055, dtype: int64

### hp: horsepower of the vehicle — a measure of its engine power.
### :There are no missing values in the hp column.
### Although hp is a real-valued numerical feature, it contains 1,055 unique values, which is relatively high but still countable.
### Horse power top Consideration:7.91 -  17754 

### 7.74  - 13936 
### 8.58  - 10111 
## 88.51 - 9968
###  8.19  -8446
####  The horse power is increasing the price also increased the customer are low


# seatCapcity

In [105]:
df["seatCapacity"].nunique()

57

In [106]:
df["seatCapacity"].value_counts()

seatCapacity
2     191919
5      85155
1      20745
7      13164
4      12523
3       8653
0       4226
8       2501
6       1386
10       638
41       278
12       178
36       164
9        153
13       150
51       104
57        88
42        83
47        77
56        77
11        76
58        75
37        62
40        61
23        59
50        50
32        39
33        34
24        30
53        30
22        29
43        28
31        24
25        24
14        22
60        22
29        21
39        20
28        18
35        16
45        15
48        14
49        13
18        13
20        12
27        12
21        11
44         9
26         8
17         6
52         6
55         4
19         4
16         2
46         2
61         2
38         2
Name: count, dtype: int64

### Seat Capacity Meaning:
The seatCapacity column refers to the number of seating positions in a vehicle, including the driver.
### Missing Values:
There are no missing values in this column. And 57 unique values
### Top 5 Most Common Seat Capacities:
### 2 seats → 191,919 vehicles
### 5 seats → 85,155 vehicles
## 1 seat → 20,745 vehicles
### 7 seats → 13,164 vehiclmes
### 4 seats → 12,523 vehicles
### Bottom 2 Least Common Seat Capacities:
### 61 seats → 2 vehicles
### 38 seats → 2 vehicles
### (others with similar low counts)



In [107]:
df["OfficeCd"].nunique()

56

In [108]:
df["OfficeCd"].value_counts()

OfficeCd
RTA RANGAREDDY              28675
RTA MEDCHAL                 22680
RTA IBRAHIMPATNAM           22148
RTA UPPAL                   20818
RTA-HYDERABAD-CZ            19897
RTA-HYDERABAD-SZ            11566
RTA-HYDERABAD-NZ            11405
RTA-HYDERABAD-WZ            11050
RTA-HYDERABAD-EZ             9286
RTA NALGONDA                 9161
RTA NIZAMABAD                8357
UNIT OFFICE KUKATPALLY       8302
RTA KHAMMAM                  7882
RTA KARIMNAGAR               7858
RTA SIDDIPET                 7311
RTA MAHABOOBNAGAR            6831
RTA YADADRI                  6341
RTA SANGAREDDY               6267
RTA HANUMAKONDA              6199
UNIT OFFICE PATANCHERUVU     6139
RTA KAMAREDDY                6122
RTA WARANGAL                 5284
RTA SURYAPET                 5265
RTA MANCHERIAL               4909
RTA BHADRADRI                4520
UNIT OFFICE MIRYALAGUDA      4381
RTA MEDAK                    4340
RTA ADILABAD                 3982
RTA NIRMAL                   3761
RTA J

### Office Code Standardization: 

### The officeCd column contains a prefix "RTA" in many entries; to ensure uniformity, we remove this prefix to retain only the actual office name (e.g., "SIDDIPET" instead of "RTA SIDDIPET").
## there are unique 56 “RTA” offices data is available and there is no missing values
.


# sumamry 

Data Exploration & Initial Assessment

### Data Volume: The dataset includes 343,137 vehicle records with detailed  16 features like registration, fuel type, and vehicle specifications.
### Datetime Conversion: Object-type columns like regvalidfrom, regvalidto, fromDate, and toDate were successfully converted to datetime format.
### Vehicle Makers & Models: ONDA, MARUTI, and BAJAJ are top manufacturers; ACTIVA-DLX BSVI-PH2 is the most common vehicle model.
### Fuel Type Issues: 13 fuel types were found, including invalid values (-1, 0, lowercase petrol), and PETROL dominates usage.
### Cylinders & cc : Vehicles commonly have 1–4 cylinders, with popular cc values like 124.0, 97.0, and 1197.0, indicating a large number of two-wheelers.
### Body Type Variety: 39 body types were found, with Solo, Sedan, and Open dominating; some rare entries appeared only once or twice.



# phases 2 project

In [109]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 343137 entries, 0 to 343136
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   slno            343137 non-null  int64  
 1   registrationNo  341450 non-null  object 
 2   reserve_no      23835 non-null   object 
 3   regvalidfrom    341445 non-null  object 
 4   regvalidto      343137 non-null  object 
 5   makerName       343137 non-null  object 
 6   modelDesc       343137 non-null  object 
 7   bodyType        343133 non-null  object 
 8   cc              343137 non-null  float64
 9   cylinder        343137 non-null  int64  
 10  fuel            339359 non-null  object 
 11  hp              343137 non-null  float64
 12  seatCapacity    343137 non-null  int64  
 13  OfficeCd        343137 non-null  object 
 14  fromdate        343137 non-null  object 
 15  todate          343137 non-null  object 
dtypes: float64(2), int64(3), object(11)
memory usage: 41.9+ 

In [110]:
mask = df["registrationNo"].isna() & df["reserve_no"].notna()
df.loc[mask, "registrationNo"] = df.loc[mask, "reserve_no"]


#### so it is fiilling missing values and satisfy above criteria

In [111]:
df.isna().sum()

slno                   0
registrationNo         1
reserve_no        319302
regvalidfrom        1692
regvalidto             0
makerName              0
modelDesc              0
bodyType               4
cc                     0
cylinder               0
fuel                3778
hp                     0
seatCapacity           0
OfficeCd               0
fromdate               0
todate                 0
dtype: int64

In [118]:
def body type():


SyntaxError: expected '(' (2238748230.py, line 1)

#### no need fill of the missing values   reserve_no

In [1]:
# filling body type

In [113]:
def fill_missing_body_type(row, df):
    if pd.isna(row["bodyType"]):
        # Find similar rows
        mask = (
            (df["seatCapacity"] == row["seatCapacity"]) &
            (df["hp"] == row["hp"]) &
            (df["cylinder"] == row["cylinder"]) &
            (df["bodyType"].notna())
        )
        candidates = df.loc[mask, "bodyType"]
        if not candidates.empty:
            return candidates.mode().iloc[0]  # Use the most common bodyType
        else:
            return row["bodyType"]  # Still NaN
    else:
        return row["bodyType"]  # Already filled

# Apply it to the DataFrame
df["bodyType"] = df.apply(lambda row: fill_missing_body_type(row, df), axis=1)


In [114]:
def fill_missing_fuel_type(row, df):
    if pd.isna(row["fuel"]):
        # Find similar rows
        mask = (
            (df["seatCapacity"] == row["seatCapacity"]) &
            (df["hp"] == row["hp"]) &
            (df["cylinder"] == row["cylinder"]) &
            (df["fuel"].notna())
        )
        candidates = df.loc[mask, "fuel"]
        if not candidates.empty:
            return candidates.mode().iloc[0]  # Use the most common fuel type
        else:
            return row["fuel"]  # Still NaN
    else:
        return row["fuel"]  # Already filled

# Apply it to the DataFrame
df["fuel"] = df.apply(lambda row: fill_missing_fuel_type(row, df), axis=1)


## outlier duplicates  treatment 

In [115]:
df.drop_duplicates(subset=["slno"],keep ="first", inplace = True)

In [116]:
df.drop_duplicates(subset=["registrationNo"],keep= "first",inplace = True)

In [117]:
len(df)

174111

### cylinder columns  outlier are un known values cylinder count =2979

In [57]:
x = df["cylinder"] == 2979
df.drop(index=df[x].index, inplace=True)


In [58]:
df["cylinder"].value_counts()

cylinder
1     94463
4     44946
3     20446
0      8681
6      3342
2      2066
5        36
8        13
12        2
10        2
Name: count, dtype: int64

In [ ]:
df.info()

In [127]:
import pandas as pd

# List of columns to convert
date_columns = ["regvalidfrom", "fromdate", "todate", "regvalidto"]

# Convert each column to datetime with format='mixed' and error handling
for col in date_columns:
    df[col] = pd.to_datetime(df[col], format='mixed', dayfirst=True)


In [128]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 174111 entries, 0 to 343136
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype                    
---  ------          --------------   -----                    
 0   slno            174111 non-null  int64                    
 1   registrationNo  174110 non-null  object                   
 2   reserve_no      11359 non-null   object                   
 3   regvalidfrom    174055 non-null  datetime64[ns]           
 4   regvalidto      174111 non-null  datetime64[ns, UTC+05:30]
 5   makerName       174111 non-null  object                   
 6   modelDesc       174111 non-null  object                   
 7   bodyType        174111 non-null  object                   
 8   cc              174111 non-null  float64                  
 9   cylinder        174111 non-null  int64                    
 10  fuel            174111 non-null  object                   
 11  hp              174111 non-null  float64                 

slno registrationNo reserve_no regvalidfrom  \
0   9287073      TS36G2115        NaN   02/03/2020   
1  10311462      TS21J6061        NaN   29/01/2021   
2  10949878     TS08HW0365        NaN   27/11/2021   
3   5255752     TS08FB0182        NaN   03/11/2016   
4   1506087     AP29AJ8075        NaN   25/11/2009   

                      regvalidto                       makerName  \
0  2035-03-01T17:12:20.747+05:30       MARUTI SUZUKI INDIA LTD.,   
1  2036-01-28T16:46:36.317+05:30       MARUTI SUZUKI INDIA LTD.,   
2  2036-11-26T16:32:35.737+05:30  HONDA MOTORCYCLE&SCOOTER(I)P L   
3   2031-11-02T16:55:19.72+05:30  HONDA MOTORCYCLE&SCOOTER(I)P L   
4  2030-01-01T11:00:52.313+05:30       MARUTI SUZUKI INDIA LTD.,   

                                         modelDesc   bodyType      cc  \
0               MARUTI XL6 SMART HYBRID ALPHA BSVI  HATCHBACK  1462.0   
1                         MARUTI S-PRESSO VXI BSVI  HATCHBACK   998.0   
2               ACTIVA 6G WITH LED HEAD LIGHT BSVI       Solo   109.0   
3  ACTIVA 3G W ELECTRIC A START&K START &CBS BSIII       Solo   109.0   
4                            MARUTI  RITZ LDI BSIV       Solo  1248.0   

   cylinder    fuel      hp  seatCapacity      OfficeCd    fromdate  \
0         4  PETROL  103.25             6  RTA SIDDIPET  01/01/2025   
1         3  PETROL   67.05             5  RTA JAGITYAL  01/01/2025   
2         1  PETROL    7.68             2     RTA UPPAL  01/01/2025   
3         1  PETROL    8.00             2     RTA UPPAL  01/01/2025   
4         4  DIESEL   55.00             5  RTA SIDDIPET  01/01/2025   

       todate  
0  02/01/2025  
1  02/01/2025  
2  02/01/2025  
3  02/01/2025  
4  02/01/2025

In [136]:
df.dropna(subset=["registrationNo","regvalidfrom"],inplace =True)

In [137]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 174054 entries, 0 to 343136
Data columns (total 16 columns):
 #   Column          Non-Null Count   Dtype                    
---  ------          --------------   -----                    
 0   slno            174054 non-null  int64                    
 1   registrationNo  174054 non-null  object                   
 2   reserve_no      11306 non-null   object                   
 3   regvalidfrom    174054 non-null  datetime64[ns]           
 4   regvalidto      174054 non-null  datetime64[ns, UTC+05:30]
 5   makerName       174054 non-null  object                   
 6   modelDesc       174054 non-null  object                   
 7   bodyType        174054 non-null  object                   
 8   cc              174054 non-null  float64                  
 9   cylinder        174054 non-null  int64                    
 10  fuel            174054 non-null  object                   
 11  hp              174054 non-null  float64                 

# without loss one column finally got the final stages